### PR TITLE
Fixed #31825 -- Made RenameField operation a noop for fields with db_column.

### PR DESCRIPTION
**Pull request by iurisilvio**
https://code.djangoproject.com/ticket/31825

### CODE REVIEW

File: tests/migrations/test_operations.py [hunk 1]
```diff
@@ -1632,6 +1632,45 @@ def test_rename_field(self):
         self.assertEqual(definition[1], [])
         self.assertEqual(definition[2], {'model_name': "Pony", 'old_name': "pink", 'new_name': "blue"})
 
+    def test_rename_field_with_db_column_defined(self):
+        """
+        RenameField with a db_column defined is a noop migration.
+        """
```
Comment (felixxm) [inline tests/migrations/test_operations.py:1638]: We can chop this docstring.

File: tests/migrations/test_operations.py [hunk 2]
```diff
@@ -1632,6 +1632,45 @@ def test_rename_field(self):
...
+        project_state = self.set_up_test_model("test_rnfl", unique_together=True, index_together=True)
```
Comment (felixxm) [inline tests/migrations/test_operations.py:1639]: I would rather create a custom model with field that has `db_column`, e.g.
```python
        project_state = self.apply_operations('test_rfwdbc', ProjectState(), operations=[
            migrations.CreateModel('Pony', fields=[
                ('id', models.AutoField(primary_key=True)),
                ('field', models.IntegerField(db_column='db_field')),
            ]),
        ])
        operation = migrations.RenameField('Pony', 'field', 'renamed_field')
        new_state = project_state.clone()
        operation.state_forwards('test_rfwdbc', new_state)
```

File: tests/migrations/test_operations.py [hunk 3]
```diff
@@ -1632,6 +1632,45 @@ def test_rename_field(self):
...
+        new_state = project_state.clone()
+        operation.state_forwards("test_rnfl", new_state)
+
+        # Model changed
+        self.assertIn("blue", new_state.models["test_rnfl", "pony"].fields)
+        self.assertNotIn("pink", new_state.models["test_rnfl", "pony"].fields)
+
+        # But database are still the same
+        self.assertColumnExists("test_rnfl_pony", "pink")
+        self.assertColumnNotExists("test_rnfl_pony", "blue")
```
Comment (felixxm) [inline tests/migrations/test_operations.py:1657]: I don't see much value in this check.

File: tests/migrations/test_operations.py [hunk 4]
```diff
@@ -1632,6 +1632,45 @@ def test_rename_field(self):
...
+
+        with connection.schema_editor() as editor:
+            # Must be a noop.
+            with self.assertNumQueries(0):
+                operation.database_backwards("test_rnfl", editor, new_state, project_state)
+
+        self.assertColumnExists("test_rnfl_pony", "pink")
+        self.assertColumnNotExists("test_rnfl_pony", "blue")
```
Comment (felixxm) [inline tests/migrations/test_operations.py:1672]: I think we can simplify this test:
```suggestion
        self.assertIn('renamed_field', new_state.models['test_rfwdbc', 'pony'].fields)
        self.assertNotIn('field', new_state.models['test_rfwdbc', 'pony'].fields)
        self.assertColumnExists('test_rfwdbc_pony', 'db_field')
        with connection.schema_editor() as editor:
            with self.assertNumQueries(0):
                operation.database_forwards('test_rfwdbc', editor, project_state, new_state)
        self.assertColumnExists('test_rfwdbc_pony', 'db_field')
        with connection.schema_editor() as editor:
            with self.assertNumQueries(0):
                operation.database_backwards('test_rfwdbc', editor, new_state, project_state)
        self.assertColumnExists('test_rfwdbc_pony', 'db_field')
```

**Comment by felixxm** [comment]
I was able to reproduce this behavior only on SQLite. I'm not sure if `RenameField` is the best place for this optimization because custom backends may rely on field names instead of db columns. Maybe we should move it to SQLite backend, e.g.
```diff
diff --git a/django/db/backends/sqlite3/schema.py b/django/db/backends/sqlite3/schema.py
index aca5d61b39..df729c8eca 100644
--- a/django/db/backends/sqlite3/schema.py
+++ b/django/db/backends/sqlite3/schema.py
@@ -102,7 +102,7 @@ class DatabaseSchemaEditor(BaseDatabaseSchemaEditor):
         old_field_name = old_field.name
         table_name = model._meta.db_table
         _, old_column_name = old_field.get_attname_column()
-        if (new_field.name != old_field_name and
+        if (new_field.column != old_field.column and
                 not self.connection.features.supports_atomic_references_rename and
                 self._is_referenced_by_fk_constraint(table_name, old_column_name, ignore_self=True)):
             if self.connection.in_atomic_block:
@@ -356,8 +356,13 @@ class DatabaseSchemaEditor(BaseDatabaseSchemaEditor):
             not (old_field.remote_field and old_field.db_constraint or
                  new_field.remote_field and new_field.db_constraint)):
             return self.execute(self._rename_field_sql(model._meta.db_table, old_field, new_field, new_type))
-        # Alter by remaking table
-        self._remake_table(model, alter_field=(old_field, new_field))
+        # Don't alter if column doesn't change.
+        if (
+            self.column_sql(model, old_field) != self.column_sql(model, new_field)
+            or old_field.column != new_field.column
+        ):
+            # Alter by remaking table
+            self._remake_table(model, alter_field=(old_field, new_field))
         # Rebuild tables with FKs pointing to this field.
         if new_field.unique and old_type != new_type:
             related_models = set()
```

**Comment by iurisilvio** [comment]
Thanks @felixxm for this great review!

I agree maybe the best to make the optimization is on `alter_field` because of custom backends, but it is not SQLite only, FKs on Postgres and probably other dbs are affected too, triggering `DROP/SET CONSTRAINT` with the same constraint name. It was the bug that my original bug.

At least I must add a test for FKs. To do it in `alter_field` I'll change `base/schema.py` too.

**Comment by iurisilvio** [comment]
I applied your code review suggestions, but I will test for FKs and try to optimize on `alter_field` instead of on `RenameField` operation.

**Comment by iurisilvio** [comment]
The `alter_field` implementation was more complicated than expected. We can't just look for field name/column, we have to check if **only** name changed. https://github.com/django/django/pull/13236/commits/f82633a37f19acece5d7ec81d999a5ba53cef23d

I'm not squashing yet, because it is a work in progress.

**Comment by iurisilvio** [comment]
Added a test to FKs and it failed with postgres and mysql, as expected. https://github.com/django/django/pull/13236/commits/e66d9660a5d8575b9424ec7cae161c289cd8ebc7

File: tests/migrations/test_operations.py [hunk 5]
```diff
@@ -1632,6 +1632,50 @@ def test_rename_field(self):
...
+        with connection.schema_editor() as editor:
+            with self.assertNumQueries(0):
+                operation.database_forwards('test_rfwdbc', editor, project_state, new_state)
+        self.assertColumnExists('test_rfwdbc_pony', 'db_field')
+        with connection.schema_editor() as editor:
+            with self.assertNumQueries(0):
+                operation.database_backwards('test_rfwdbc', editor, new_state, project_state)
+        self.assertColumnExists('test_rfwdbc_pony', 'db_field')
+
+    def test_rename_field_with_fk_db_column(self):
```
Comment (iurisilvio) [inline tests/migrations/test_operations.py:1657]: Both tests are almost the same. I could do it with just one test and two columns.

I don't know the "Django best practice" here.

Comment (felixxm) [reply tests/migrations/test_operations.py:1657]: We're just checking the number of queries, so a single test with two columns sounds fine.

**Comment by iurisilvio** [comment]
Tests passed, just cleaning the code. 42ec62b

**Comment by iurisilvio** [comment]
Implemented change on `alter_field` instead of `RenameField` and ready for a new code review. :)

Also tested for the foreign key case, that failed as expected with Postgres and MySQL.

File: django/db/backends/base/schema.py [hunk 6]
```diff
@@ -538,6 +538,14 @@ def alter_field(self, model, old_field, new_field, strict=False):
         If `strict` is True, raise errors if the old column does not match
         `old_field` precisely.
         """
+        # Don't alter if only field name has changed.
+        if (
+            old_field.name != new_field.name and
```
Comment (felixxm) [inline django/db/backends/base/schema.py:543]: Do we need to check this? We would also like to omit altering if `name` is the same.

File: django/db/backends/base/schema.py [hunk 7]
```diff
@@ -538,6 +538,14 @@ def alter_field(self, model, old_field, new_field, strict=False):
...
+            old_field.column == new_field.column and
+            old_field.deconstruct()[1:] == new_field.deconstruct()[1:]
+        ):
+            return
```
Comment (felixxm) [inline django/db/backends/base/schema.py:547]: I think we should create a hook similar to the `_field_should_be_indexed()`, that will allow 3rd-party backend to adjust this behavior, e.g.
```python
    def _field_should_be_altered(self, old_field, new_field):
        # Don't alter when changing only a field name.
        return (
            old_field.column != new_field.column or
            old_field.deconstruct()[1:] != new_field.deconstruct()[1:]
        )
```

File: django/db/backends/sqlite3/schema.py [hunk 8]
```diff
@@ -99,6 +99,14 @@ def alter_db_table(self, model, old_db_table, new_db_table, disable_constraints=
             super().alter_db_table(model, old_db_table, new_db_table)
 
     def alter_field(self, model, old_field, new_field, strict=False):
+        # Don't alter if only field name has changed.
+        if (
+            old_field.name != new_field.name and
+            old_field.column == new_field.column and
+            old_field.deconstruct()[1:] == new_field.deconstruct()[1:]
+        ):
```
Comment (felixxm) [inline django/db/backends/sqlite3/schema.py:107]: We could reuse `_field_should_be_altered()`:
```suggestion
        if not self._field_should_be_altered(old_field, new_field):
```

File: django/db/backends/base/schema.py (hunk 7 above)
Comment (iurisilvio) [reply django/db/backends/base/schema.py:542]: Done, it is cleaner. :+1:

**Comment by felixxm** [comment]
@iurisilvio Thanks for updates :+1:

**Comment by felixxm** [comment]
buildbot, test on oracle.

**Pull request merged**